tools/sil-llvm-gen/swift_lto_main.cpp

```diff
@@ -0,0 +1,90 @@
+///
+//===----------------------------------------------------------------------===//
+
+#include "swift/Basic/LLVMInitialize.h"
+#include "swift/Frontend/Frontend.h"
+#include "swift/LTO/LTO.h"
+#include "swift/Strings.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include <cstdio>
+#include <memory>
+using namespace swift;
+
+llvm::cl::SubCommand LTOSubcommand("lto", "Swift LTO Tool");
+
+static llvm::cl::list<std::string>
+    InputFilenames(llvm::cl::Positional, llvm::cl::desc("[input files...]"),
+                   llvm::cl::OneOrMore, llvm::cl::sub(LTOSubcommand));
+
+// This function isn't referenced outside its translation unit, but it
+// can't use the "static" keyword because its address is used for
+// getMainExecutable (since some platforms don't support taking the
+// address of main, and some platforms can't implement getMainExecutable
+// without being given the address of a function in the main executable).
+void anchorForGetMainExecutableInSwiftLTO() {}
+
+// This tool is combined with sil-llvm-gen to reduce link time.
+// This entrypoint is invoked from SILLLVMGen.cpp when user invoke
+// lto subcommand.
+int swift_lto_main(int argc, char **argv) {
+  CompilerInvocation Invocation;
+
+  Invocation.setMainExecutablePath(llvm::sys::fs::getMainExecutable(
+      argv[0],
+      reinterpret_cast<void *>(&anchorForGetMainExecutableInSwiftLTO)));
+
+  auto SearchPathOpts = Invocation.getSearchPathOptions();
+  llvm::SmallVector<llvm::StringRef, 4> RuntimeLibraryPaths;
+  llvm::SmallVector<llvm::StringRef, 4> RuntimeLibraryImportPaths;
+  RuntimeLibraryPaths.insert(RuntimeLibraryPaths.begin(),
+                             SearchPathOpts.RuntimeLibraryPaths.begin(),
+                             SearchPathOpts.RuntimeLibraryPaths.end());
+  RuntimeLibraryImportPaths.insert(
+      RuntimeLibraryImportPaths.begin(),
+      SearchPathOpts.RuntimeLibraryImportPaths.begin(),
+      SearchPathOpts.RuntimeLibraryImportPaths.end());
+  lto::LTOPipeline Pipeline(RuntimeLibraryPaths, RuntimeLibraryImportPaths,
+                            SearchPathOpts.RuntimeResourcePath);
+
+  for (auto InputFilename : InputFilenames) {
+    // Load the input file.
+    auto FileBufOrErr = llvm::MemoryBuffer::getFileOrSTDIN(InputFilename);
+    if (!FileBufOrErr) {
+      fprintf(stderr, "Error! Failed to open file: %s\n",
+              InputFilename.c_str());
+      exit(-1);
+    }
+
+    if (Pipeline.addModule(std::move(FileBufOrErr.get()))) {
+      fprintf(stderr, "Error! Failed to load serialized module: %s\n",
+              InputFilename.c_str());
+      exit(-1);
+    }
+  }
+
+  Pipeline.emitLLVMModules([&](StringRef ModuleName) {
+    std::error_code EC;
+    std::unique_ptr<llvm::raw_ostream> RawOS =
+        std::make_unique<llvm::raw_fd_ostream>(ModuleName.str() + ".bc", EC);
+    if (EC)
+      return std::unique_ptr<llvm::raw_ostream>(nullptr);
+    return RawOS;
+  });
```

Comment:
Can you clarify why exact this tool is needed?  This really feels like its executing `sil-opt` + `swift-llvm-gen`.  Is there a reason that you cannot just pipe the tools in the test?

Reply:
The tool accepts SIB files for multiple modules. This is something swift-llvm-gen does not do.

Reply:
But you can load the SIB file in `sil-opt`, run it through the pipeline and write that out, then use `swift-llvm-gen` to generate IR from the SIB.

> If piping sil-opt and swift-llvm-gen, it's impossible to optimize across modules.
I think Im not understanding the missing piece in the tools.  Can you explain that and write that up in the commit message as well?  I think its critical to explain the deficiency.

> And the main purpose of this tool is simulating linker integration without linker. So it doesn't match sil-opt and swift-llvm-gen's purposes.

I disagree on this.  Its not simulating linker integration - its running the SIL pipeline on a set of SIB files.  That is not related to linking.  Running the SIL pipeline is precisely what `sil-opt` does.  You seem to generate LLVM IR from the binary at the end, and thats precisely what `swift-llvm-gen` does - generates LLVM IR from SIL/SIB.  That is where the disconnect is, and what Im trying to understand.

Reply:
As you say, this swift-lto and sil-opt have the same part of roles in optimizing SIBs. But the sil-opt also accept SIL as input, whereas the LTO pipeline only accepts SIBs.

The LTO pipeline accepts inputs only from linkers, so SILs will never be inputted.

If LTO pipeline accepts SILs as input, the LTO pipeline becomes more complex only for testing.

For example, SILs doesn't serialize SDKPath, etc., so we have to assemble those missing pieces of information before passing input them into the pipeline.

So creating new tool that only accepts SIBs and  only helps keep the tool simple.

And sil-opt really depends on Frontend module that assumes that input module is always one. But LTO pipeline is independent from the module because LTO pipeline needs to be acceptable for multi-modules. This difference makes it difficult to merge these tools.

Reply:
Right, `sil-opt` accepts SIL and SIB, but that makes it more flexible and keeps everything in one place :).

The flags are precisely what all the LTO work is about - its how to thread all the appropriate flags through the driver into the frontend and the linker.  I think that is expected.  That will also be needed for this tool.

So given that the tools can accomplish what is needed, the question is why is the new tool better than building on the existing tools?  Is it going to be far more code to just add support for multiple files to `sil-opt` than writing a new tool?

Reply:
Perhaps I've made a big misunderstanding. Do you mean that the linker will run sil-opt directly?

I was thinking that LTOPipeline class will be linked to the linker. And swift-lto The tool was only going to be used for testing.